roles/rsyslog/tasks/main.yml

```diff
@@ -2,8 +2,7 @@
   package:
     name: "{{ (__rsyslog_base_packages) + (__rsyslog_tls_packages if rsyslog_pki else []) + (rsyslog_extra_packages | flatten) }}"
     state: latest
-  when:
-    - rsyslog_enabled | bool
+  when: __rsyslog_enabled
```

Comment:
doesn't this lead to a warning about bare variable?

Reply:
yes - it does - so I guess the moral of the story is that you cannot use a bare variable in a `when` - if it is a boolean, you can use `when: varname | bool` - if it is some other type, I'm not sure what to do.

I created a test playbook
```yaml
---
- name: set parameters for test
  hosts: localhost
  tasks:
  - set_fact:
      junkdict: {'foo':'bar'}

  - debug:
      msg: hereherehere
    when: junkdict
```
running this - `ansible-playbook junk-playbook.yml` - gives
```
TASK [debug] *******************************************************************
[DEPRECATION WARNING]: evaluating {'foo': 'bar'} as a bare variable, this 
behaviour will go away and you might need to add |bool to the expression in the
 future. Also see CONDITIONAL_BARE_VARS configuration toggle.. This feature 
will be removed in version 2.12. Deprecation warnings can be disabled by 
setting deprecation_warnings=False in ansible.cfg.
ok: [localhost] => {
    "msg": "hereherehere"
}
```
so it "works" - that is, it correctly evaluates the `dict` value as non-empty - but we get the warning - so I tried using `when: junkdict|bool` and I get this:
```
TASK [debug] *******************************************************************
skipping: [localhost]
```
That doesn't work - if you filter a dict as a bool, it does not return a `true` value.  I tried this too - `when: "{{ junkdict }}"`:
```
TASK [debug] *******************************************************************
[WARNING]: conditional statements should not include jinja2 templating
delimiters such as {{ }} or {% %}. Found: {{ junkdict }}
ok: [localhost] => {
    "msg": "hereherehere"
}
```
it also "works" but with a warning.  The only thing I have found that works consistently without a warning is this: `when: junkdict | length > 0` - but that only works with types that have a length (string, dict, list) - also note that with types of `int`, you can use `when: var|bool` but note that `0` evaluates to `false`, and *all other values* evaluate to `true` which may or may not be what you want

Reply:
@richm the problem is, `| bool` is not a conversion to boolean in the Python sense. It is a conversion of string like `"yes"` -> true, `"no"` -> false. For a conversion in the Python sense this you would just use the variable and Jinja2 would convert it for you, except that in case it is just a variable and not a more complex expression you hit the "bare variable" rule (which is fortunately being deprecated). Plenty of discussions of this can be found in related Ansible issues. Didn't I say that this is confusing?

Reply:
> doesn't this lead to a warning about bare variable? Also for the cases below.

@pcahyna, do we need to wrap variables in the when conditional?
https://docs.ansible.com/ansible/latest/user_guide/playbooks_conditionals.html
According to this user guide,
`As the examples show, you don’t need to use {{ }} to use variables inside conditionals, as these are already implied.`
In our case, __rsyslog_enabled is considered "already implied" or not?  Anyway, I don't see any warning messages in terms of __rsyslog_enabled.

Reply:
I think you can work around by using one of these which should be equivalent and not behave like bare variables:
```
not (not junkdict)
junkdict | d()
( junkdict )
```
I have not tested them though

Reply:
or
```
vars.junkdict
```

Reply:
> > doesn't this lead to a warning about bare variable? Also for the cases below.
> 
> @pcahyna, do we need to wrap variables in the when conditional?
> https://docs.ansible.com/ansible/latest/user_guide/playbooks_conditionals.html

No.

> According to this user guide,
> `As the examples show, you don’t need to use {{ }} to use variables inside conditionals, as these are already implied.`
> In our case, __rsyslog_enabled is considered "already implied" or not? Anyway, I don't see any warning messages in terms of __rsyslog_enabled.

@richm says that it does lead to warning messages.

Reply:
Anyway, I would say that we should not do this change yet, because it is not clear what the best practice is.

Reply:
related discussion: https://github.com/linux-system-roles/logging/pull/87#pullrequestreview-364901931

Reply:
> Anyway, I would say that we should not do this change yet, because it is not clear what the best practice is.

Ok, I'm going to put the bool filter back.  Thanks.